packages/hash/shared/src/types.ts

```diff
@@ -0,0 +1,339 @@
+    description: "An emoji icon.",
+  },
+  tokens: {
+    title: "Tokens",
+    description: undefined,
+  },
+} as const;
+
+type WorkspacePropertyTypeKey = keyof typeof workspacePropertyTypes;
+
+export type WorkspacePropertyTypeTitle =
+  typeof workspacePropertyTypes[WorkspacePropertyTypeKey]["title"];
+
+/**
+ * The workspace link type titles.
+ */
+const workspaceLinkTypes = {
+  hasMembership: {
+    title: "Has Membership",
+    description: "Having a membership.",
+  },
+  ofOrg: {
+    title: "Of Org",
+    description: "Belonging to an organization",
+  },
+  blockData: {
+    title: "Block Data",
+    description: "The entity representing the data in a block.",
+  },
+  contains: {
+    title: "Contains",
+    description: "Something containing something.",
+  },
+  parent: {
+    title: "Parent",
+    description: "The parent of something.",
+  },
+} as const;
+
+type WorkspaceLinkTypeKey = keyof typeof workspaceLinkTypes;
+
+export type WorkspaceLinkTypeTitle =
+  typeof workspaceLinkTypes[WorkspaceLinkTypeKey];
+
+/**
+ * The primitive data types ("Text", "Number", etc.)
+ */
+const primitiveDataTypes = {
+  text: {
+    title: "Text",
+    description: "An ordered sequence of characters",
+  },
+  number: {
+    title: "Number",
+    description: "An arithmetical value (in the Real number system)",
+  },
+  boolean: {
+    title: "Boolean",
+    description: "A True or False value",
+  },
+  emptyList: {
+    title: "Empty List",
+    description: "An Empty List",
+  },
+  object: {
+    title: "Object",
+    description: "A plain JSON object with no pre-defined structure",
+  },
+  null: {
+    title: "Null",
+    description: "A placeholder value representing 'nothing'",
+  },
+} as const;
+
+export type PrimitiveDataTypeKey = keyof typeof primitiveDataTypes;
+
+export type PrimitiveDataTypeTitle =
+  typeof primitiveDataTypes[PrimitiveDataTypeKey]["title"];
+
+type TypeDefinition = {
+  title: string;
+  description?: string;
+};
+
+type EntityTypeDefinition = TypeDefinition & { entityTypeId: VersionedUri };
+
+type PropertyTypeDefinition = TypeDefinition & { propertyTypeId: VersionedUri };
+
+type DataTypeDefinition = TypeDefinition & { dataTypeId: VersionedUri };
+
+type LinkTypeDefinition = TypeDefinition & {
+  linkTypeId: VersionedUri;
+  description: string;
+};
+
+type TypeDefinitions = {
+  entityType: Record<WorkspaceEntityTypeKey, EntityTypeDefinition>;
+  propertyType: Record<WorkspacePropertyTypeKey, PropertyTypeDefinition>;
+  linkType: Record<WorkspaceLinkTypeKey, LinkTypeDefinition>;
+  dataType: Record<PrimitiveDataTypeKey, DataTypeDefinition>;
+};
+
+/**
+ * The workspace and block protocol types that are statically available at run-time.
+ *
+ * @todo - instead of just providing the id, title and description of types, find a way
+ * to make the entire schemas statically available at run-time.
+ */
+export const types: TypeDefinitions = {
+  entityType: Object.entries(workspaceEntityTypes).reduce(
```

Comment:
This whole file feels kind of unwieldy. It doesn't seem like it should be this complicated, (using reduces here and loads of intermediary objects for example feels quite.. 'clever') but I don't have any good suggestions on how to simplify it at the moment without being incredibly verbose.

Just want to flag the concern in case anyone else has more inspiration today, but otherwise we can go ahead and hope we clean it up in the near future.

Reply:
I agree. A lot of the complexity is caused by the fact that the identifier is generated by the `title`, and both need to be made available statically. I've tried to minimise duplicate code through the reduce methods, but appreciate they're difficult to digest. Don't have a better idea in mind either, so I think we should leave this for now unless someone has a more elegant solution that doesn't introduce duplicate method calls throughout this file.